File: schedulerApi/models/sites/workcenter.go
package sites

import "go.mongodb.org/mongo-driver/bson/primitive"

type Shift struct {
	ID              string   `json:"id" bson:"id"`
	Name            string   `json:"name" bson:"name"`
	SortID          uint     `json:"sortid" bson:"sortid"`
	AssociatedCodes []string `json:"associatedCodes,omitempty" bson:"associatedCodes,omitempty"`
	PayCode         uint     `json:"payCode" bson:"payCode"`
}

type ByShift []Shift

func (c ByShift) Len() int { return len(c) }
func (c ByShift) Less(i, j int) bool {
	return c[i].SortID < c[j].SortID
}
func (c ByShift) Swap(i, j int) { c[i], c[j] = c[j], c[i] }

type Position struct {
	ID       string               `json:"id" bson:"id"`
	Name     string               `json:"name" bson:"name"`
	SortID   uint                 `json:"sortid" bson:"sortid"`
	Assigned []primitive.ObjectID `json:"assigned" bson:"assigned"`
}

type ByPosition []Position

func (c ByPosition) Len() int { return len(c) }
func (c ByPosition) Less(i, j int) bool {
	return c[i].SortID < c[j].SortID
}
func (c ByPosition) Swap(i, j int) { c[i], c[j] = c[j], c[i] }

type Workcenter struct {
	ID        string     `json:"id" bson:"id"`
	Name      string     `json:"name" bson:"name"`
	SortID    uint       `json:"sortid" bson:"sortid"`
	Shifts    []Shift    `json:"shifts,omitempty" bson:"shifts,omitempty"`
	Positions []Position `json:"positions,omitempty" bson:"positions,omitempty"`
}

type ByWorkcenter []Workcenter

func (c ByWorkcenter) Len() int { return len(c) }
func (c ByWorkcenter) Less(i, j int) bool {
	return c[i].SortID < c[j].SortID
}
func (c ByWorkcenter) Swap(i, j int) { c[i], c[j] = c[j], c[i] }

File: scheduler-client/src/app/models/sites/workcenter.ts
export interface IShift {
  id: string;
  name: string;
  sort: number;
  associatedCodes?: string[];
  payCode: number;
}

export class Shift implements IShift {
  id: string;
  name: string;
  sort: number;
  associatedCodes?: string[];
  payCode: number;

  constructor(shft?: IShift) {
    this.id = (shft) ? shft.id : '';
    this.name = (shft) ? shft.name : '';
    this.sort = (shft) ? shft.sort : 0;
    this.payCode = (shft) ? shft.payCode : 1;
    this.associatedCodes = [];
    if (shft && shft.associatedCodes) {
      shft.associatedCodes.forEach(ac => {
        this.associatedCodes?.push(ac);
      });
    }
  }

  compareTo(other?: Shift): number {
    if (other) {
      return (this.sort < other.sort) ? -1 : 1;
    }
    return -1;
  }
}

export interface IPosition {
  id: string;
  name: string;
  sort: number;
  assigned: string[];
}

export class Position implements IPosition {
  id: string;
  name: string;
  sort: number;
  assigned: string[];

  constructor(pos?: IPosition) {
    this.id = (pos) ? pos.id : '';
    this.name = (pos) ? pos.name : '';
    this.sort = (pos) ? pos.sort : 0;
    this.assigned = [];
    if (pos && pos.assigned.length > 0) {
      pos.assigned.forEach(asgn => {
        this.assigned.push(asgn);
      });
    }
  }

  compareTo(other?: Position): number {
    if (other) {
      return (this.sort < other.sort) ? -1 : 1;
    }
    return -1;
  }
}

export interface IWorkcenter {
  id: string;
  name: string;
  sort: number;
  shifts?: IShift[];
  positions?: IPosition[];
}

export class Workcenter implements IWorkcenter {
  id: string;
  name: string;
  sort: number;
  shifts?: Shift[];
  positions?: Position[];

  constructor(wc?: IWorkcenter) {
    this.id = (wc) ? wc.id : '';
    this.name = (wc) ? wc.name : '';
    this.sort = (wc) ? wc.sort : 0;
    this.shifts = [];
    if (wc && wc.shifts && wc.shifts.length > 0) {
      wc.shifts.forEach(s => {
        this.shifts?.push(new Shift(s));
      });
      this.shifts?.sort((a,b) => a.compareTo(b))
    }
    this.positions = [];
    if (wc && wc.positions && wc.positions.length > 0) {
      wc.positions.forEach(pos => {
        this.positions?.push(new Position(pos));
      });
      this.positions.sort((a,b) => a.compareTo(b));
    }
  }
}